File: iosApp/iosApp/Screens/Settings/SettingsLogic.swift
import Foundation
import Shared

/// Pure derivations for the Settings, Advanced, and Balance screens (U22,
/// R12, R14): the PWA's row tables, the appearance radiogroup mapping, and
/// the balance breakdown — screens only place the results. Ported
/// value-for-value from Android's `SettingsLogic.kt`; copy is verbatim from
/// `Settings.tsx`, `Advanced.tsx`, and `Balance.tsx`.

/// One tappable settings row; `destination` nil = inert no-op (PWA parity).
struct SettingsRowSpec: Equatable {
    let label: String
    let detail: String
    let destination: Route?
}

/// `SETTINGS_ITEMS` (`Settings.tsx:12-107`). How It Works and Get Help ship
/// as no-ops in the PWA (`route: null`) — replicated as inert rows (plan
/// Scope Boundaries).
let settingsRows: [SettingsRowSpec] = [
    SettingsRowSpec(label: "Wallet Backup", detail: "Setup", destination: .settingsBackup),
    SettingsRowSpec(label: "Recover Wallet", detail: "From Seed", destination: .settingsRestore),
    SettingsRowSpec(label: "Advanced", detail: "Settings", destination: .settingsAdvanced),
    SettingsRowSpec(label: "How It Works", detail: "FAQ", destination: nil),
    SettingsRowSpec(label: "Get Help", detail: "Chat with us", destination: nil),
]

/// `ADVANCED_ITEMS` (`Advanced.tsx:6-47`).
let advancedRows: [SettingsRowSpec] = [
    SettingsRowSpec(label: "Balance", detail: "Onchain · Lightning", destination: .advancedBalance),
    SettingsRowSpec(label: "Peers", detail: "Connected", destination: .advancedPeers),
]

/// The radiogroup's order: `THEME_MODES` (`theme.ts:3`) — hybrid, light, dark.
let appearanceModes: [AppearanceMode] = [.hybrid, .light, .dark]

/// `THEME_LABELS` (`Settings.tsx:6-10`).
func appearanceLabel(_ mode: AppearanceMode) -> String {
    switch mode {
    case .hybrid: return "Hybrid"
    case .light: return "Light"
    case .dark: return "Dark"
    }
}

/// What the Balance screen renders, per `use-unified-balance.ts`.
struct BalanceBreakdown: Equatable {
    /// Full on-chain (confirmed + all pending) + floored Lightning sats.
    let totalSats: Int64
    /// The `+₿X pending` line: unconfirmed external receives only.
    let pendingSats: Int64
    let onchainSats: Int64
    let lightningSats: Int64
}

func balanceBreakdown(_ balances: Balances) -> BalanceBreakdown {
    let onchain = Int64(bitPattern: balances.onchainTotalSats)
    let lightning = FormatKt.msatToSatFloor(msat: Int64(bitPattern: balances.lightningMsat))
    return BalanceBreakdown(
        totalSats: onchain + lightning,
        pendingSats: Int64(bitPattern: balances.onchainUntrustedPendingSats),
        onchainSats: onchain,
        lightningSats: lightning
    )
}

File: androidApp/src/main/kotlin/zinqq/app/screens/settings/SettingsLogic.kt
package zinqq.app.screens.settings

import uniffi.wallet_core.Balances
import zinqq.app.nav.Route
import zinqq.app.theme.AppearanceMode
import zinqq.spike.msatToSatFloor

/**
 * Pure derivations for the Settings, Advanced, and Balance screens (U17,
 * R12, R14): the PWA's row tables, the appearance radiogroup mapping, and
 * the balance breakdown — screens only place the results. Copy is verbatim
 * from `Settings.tsx`, `Advanced.tsx`, and `Balance.tsx`.
 */

/** One tappable settings row; [destination] `null` = inert no-op (PWA parity). */
data class SettingsRowSpec(
    val label: String,
    val detail: String,
    val destination: Route?,
)

/**
 * `SETTINGS_ITEMS` (`Settings.tsx:12-107`). How It Works and Get Help ship
 * as no-ops in the PWA (`route: null`) — replicated as inert rows (plan
 * Scope Boundaries).
 */
val SETTINGS_ROWS: List<SettingsRowSpec> = listOf(
    SettingsRowSpec("Wallet Backup", "Setup", Route.SettingsBackup),
    SettingsRowSpec("Recover Wallet", "From Seed", Route.SettingsRestore),
    SettingsRowSpec("Advanced", "Settings", Route.SettingsAdvanced),
    SettingsRowSpec("How It Works", "FAQ", null),
    SettingsRowSpec("Get Help", "Chat with us", null),
)

/** `ADVANCED_ITEMS` (`Advanced.tsx:6-47`). */
val ADVANCED_ROWS: List<SettingsRowSpec> = listOf(
    SettingsRowSpec("Balance", "Onchain · Lightning", Route.AdvancedBalance),
    SettingsRowSpec("Peers", "Connected", Route.AdvancedPeers),
)

/** The radiogroup's order: `THEME_MODES` (`theme.ts:3`) — hybrid, light, dark. */
val APPEARANCE_MODES: List<AppearanceMode> = listOf(
    AppearanceMode.HYBRID,
    AppearanceMode.LIGHT,
    AppearanceMode.DARK,
)

/** `THEME_LABELS` (`Settings.tsx:6-10`). */
fun appearanceLabel(mode: AppearanceMode): String = when (mode) {
    AppearanceMode.HYBRID -> "Hybrid"
    AppearanceMode.LIGHT -> "Light"
    AppearanceMode.DARK -> "Dark"
}

/** What the Balance screen renders, per `use-unified-balance.ts`. */
data class BalanceBreakdown(
    /** Full on-chain (confirmed + all pending) + floored Lightning sats. */
    val totalSats: Long,
    /** The `+₿X pending` line: unconfirmed external receives only. */
    val pendingSats: Long,
    val onchainSats: Long,
    val lightningSats: Long,
)

fun balanceBreakdown(balances: Balances): BalanceBreakdown {
    val onchain = balances.onchainTotalSats.toLong()
    val lightning = msatToSatFloor(balances.lightningMsat.toLong())
    return BalanceBreakdown(
        totalSats = onchain + lightning,
        pendingSats = balances.onchainUntrustedPendingSats.toLong(),
        onchainSats = onchain,
        lightningSats = lightning,
    )
}
